### PR TITLE
Make sure Thundra not interrupts user's handler

### PR DESCRIPTION
**Pull request by hamitb**

### CODE REVIEW

File: thundra/thundra_agent.py [hunk 1]
```diff
@@ -77,39 +77,71 @@ def __call__(self, original_func):
...
+            
+            # Invoke user handler
+            if before_done:
+                try:
+                    response = original_func(event, context)
+                    self.plugin_context['response'] = response
+                except Exception as e:
+                    try:
+                        self.plugin_context['error'] = e
+                        application_support.parse_application_tags()
```
Comment (serkan-ozal) [inline thundra/thundra_agent.py]: Application tags are independent from the invocation itself. I don't understand why we parse and clear them for every invocation.

File: thundra/thundra_agent.py [hunk 2]
```diff
@@ -77,39 +77,71 @@ def __call__(self, original_func):
...
+                    raise e
+                finally:
+                    self.stop_timer()
+            else:
+                self.stop_timer()
+                self.reporter.reports.clear()
+                return original_func(event, context)
+            
+            # After having run the user's handler
+            if before_done and not after_done:
```
Comment (serkan-ozal) [inline thundra/thundra_agent.py]: If the `prepare_and_send_reports()` fails above, `after_done` will be `False` (`before_done` is `True` as pre-hooks are triggered successfully). In this case, we are calling `prepare_and_send_reports()` again that we should not do.

Comment (hamitb) [reply thundra/thundra_agent.py]: Regardless of  the failure of `prepare_and_send_reports()` above, we are raising the user's error in line 127. So if the `prepare_and_send_reports()` fails above program execution won't even come to line 136. Am I Wrong?

Comment (serkan-ozal) [reply thundra/thundra_agent.py]: ok, got it :+1:

**Pull request merged**